core/lib/spree/core/controller_helpers/order.rb

```diff
@@ -58,7 +58,18 @@ def associate_user
           session[:guest_token] = nil
         end
 
+        def remove_invalid_session_order_id
+          return unless session[:order_id]
+
+          user = try_spree_current_user
+          order = Spree::Order.find(session[:order_id]) rescue nil
+
+          session[:order_id] = nil if order.nil? || order.user != user
```

Comment:
Actually, I think that's yet another case when we *need* to delete `session[:order_id]`. If you're a guest and your `session[:order_id]` points to an order you can't access, we should not allow it to be set as `current_order` or you will get errors when you try to add a line item.